File: src/main/java/icu/takeneko/omms/crystal/rcon/BufferHelper.java
package icu.takeneko.omms.crystal.rcon;

import java.nio.charset.StandardCharsets;

public class BufferHelper {
    public static String getString(byte[] buf, int i, int j) {
        int k = j - 1;
        int l;
        for(l = Math.min(i, k); 0 != buf[l] && l < k; ++l) {}
        return new String(buf, i, l - i, StandardCharsets.UTF_8);
    }

    public static int getIntLE(byte[] buf, int start) {
        return getIntLE(buf, start, buf.length);
    }

    public static int getIntLE(byte[] buf, int start, int limit) {
        return 0 > limit - start - 4 ? 0 : buf[start + 3] << 24 | (buf[start + 2] & 255) << 16 | (buf[start + 1] & 255) << 8 | buf[start] & 255;
    }
}

File: src/main/kotlin/icu/takeneko/omms/crystal/rcon/BufferHelper.kt
package icu.takeneko.omms.crystal.rcon

import java.nio.charset.StandardCharsets
import kotlin.math.min

object BufferHelper {
    fun getString(buf: ByteArray, i: Int, j: Int): String {
        val k = j - 1
        var l: Int = min(i, k)
        while (0 != buf[l].toInt() && l < k) {
            ++l
        }
        return String(buf, i, l - i, StandardCharsets.UTF_8)
    }

    fun getIntLE(buf: ByteArray, start: Int): Int = getIntLE(buf, start, buf.size)

    @Suppress("UnnecessaryParentheses")
    fun getIntLE(buf: ByteArray, start: Int, limit: Int): Int =
        if (0 > limit - start - 4) {
            0
        } else {
            buf[start + 3].toInt() shl 24 or
                ((buf[start + 2].toInt() and 255) shl 16) or
                ((buf[start + 1].toInt() and 255) shl 8) or
                (buf[start].toInt() and 255)
        }
}
